# Walmart Sales Prediction

## Background
Walmart faces the challenge of accurately forecasting product sales across its stores to optimize inventory management, reduce stockouts and overstock situations, and improve overall supply chain efficiency in a highly dynamic retail environment influenced by factors such as seasonality, promotions, holidays, and regional demand variations; as a Data Scientist, you are tasked with developing a robust predictive model that leverages historical sales data along with relevant features such as time-based variables, store and product attributes to generate reliable sales forecasts that can support data-driven decision-making, enhance operational planning, and ultimately improve profitability and customer satisfaction.


## About the dataset
The Walmart sales dataset is a widely used retail dataset designed for time-series forecasting and regression tasks, containing historical sales records from multiple Walmart stores across different regions. It typically includes variables such as store identifiers, department or product categories, weekly sales, and time-related features like date, along with additional attributes that capture external influences such as holidays, temperature, fuel prices, consumer price index (CPI), and unemployment rates. The dataset is structured to reflect real-world retail dynamics, where sales are affected not only by internal store operations but also by macroeconomic and seasonal factors.




In [1]:
# Import the essential packages
import numpy as np
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, r2_score, root_mean_squared_error

import matplotlib.pyplot as plt


## Load the Dataset

In [2]:
df = pd.read_csv("Walmart_Sales.csv", sep=';')

## EDA

- **Glimpse into the data**

In [3]:
df.head()

,Store,Date,Weekly_Sales,Holiday_Flag,Temperature,Fuel_Price,CPI,Unemployment
0,1,05-02-2010,1643690.90,0,42.31,2.572,211.096358,8.106
1,1,12-02-2010,1641957.44,1,38.51,2.548,211.242170,8.106
2,1,19-02-2010,1611968.17,0,39.93,2.514,211.289143,8.106
3,1,26-02-2010,1409727.59,0,46.63,2.561,211.319643,8.106
4,1,05-03-2010,1554806.68,0,46.50,2.625,211.350143,8.106


- Weekly sales: target

In [4]:
df.info() # get the metadata

<class 'pandas.DataFrame'>
RangeIndex: 6435 entries, 0 to 6434
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Store         6435 non-null   int64  
 1   Date          6435 non-null   str    
 2   Weekly_Sales  6435 non-null   float64
 3   Holiday_Flag  6435 non-null   int64  
 4   Temperature   6435 non-null   float64
 5   Fuel_Price    6435 non-null   float64
 6   CPI           6435 non-null   float64
 7   Unemployment  6435 non-null   float64
dtypes: float64(5), int64(2), str(1)
memory usage: 465.2 KB


- 6435 rows
- Weekly sales: target
- no nulls

In [5]:
df.duplicated().sum() # Check for duplicates

np.int64(0)

- Weekly sales: target
- 6435 rows
- no nulls
- no duplicates

- **Get summary stats**

In [6]:
df.describe()

,Store,Weekly_Sales,Holiday_Flag,Temperature,Fuel_Price,CPI,Unemployment
count,6435.000000,6.435000e+03,6435.000000,6435.000000,6435.000000,6435.000000,6435.000000
mean,23.000000,1.046965e+06,0.069930,60.663782,3.358607,171.578394,7.999151
std,12.988182,5.643666e+05,0.255049,18.444933,0.459020,39.356712,1.875885
min,1.000000,2.099862e+05,0.000000,-2.060000,2.472000,126.064000,3.879000
25%,12.000000,5.533501e+05,0.000000,47.460000,2.933000,131.735000,6.891000
50%,23.000000,9.607460e+05,0.000000,62.670000,3.445000,182.616521,7.874000
75%,34.000000,1.420159e+06,0.000000,74.940000,3.735000,212.743293,8.622000
max,45.000000,3.818686e+06,1.000000,100.140000,4.468000,227.232807,14.313000


# Model training

- **Extract features and targets**

In [7]:
X = df.drop(["Date", "Weekly_Sales"], axis=1)
y = df["Weekly_Sales"]

- **Split the dataset for training and testing**

In [8]:
X_train, X_test, y_train, y_test =  train_test_split(X, y, test_size=0.2, random_state=30)

- **Initialise the regressor**

In [9]:
random_forest = RandomForestRegressor(n_estimators=100, max_depth=10, random_state=39)

- **Train the model**

In [10]:
model = random_forest.fit(X_train, y_train)

- **Make predictions**

In [12]:
y_pred = np.round(model.predict(X_test),2)

- **Evaluate the performance of the model**

In [13]:
eval_dic = {"predicted": y_pred, "true": y_test}
eval_df = pd.DataFrame(eval_dic)
eval_df

,predicted,true
2593,1454397.53,1457314.39
6415,776629.40,821498.18
553,2159429.23,2197299.65
4768,1086052.39,898610.33
1883,2153338.39,2044155.39
...,...,...
2128,622148.31,678024.75
5389,416301.75,419717.41
100,1601827.44,1550369.92
4884,1160744.12,1268766.76


In [14]:
eval_df["error"] = np.round(eval_df['predicted'] - eval_df['true'], 2)

In [15]:
eval_df

,predicted,true,error
2593,1454397.53,1457314.39,-2916.86
6415,776629.40,821498.18,-44868.78
553,2159429.23,2197299.65,-37870.42
4768,1086052.39,898610.33,187442.06
1883,2153338.39,2044155.39,109183.00
...,...,...,...
2128,622148.31,678024.75,-55876.44
5389,416301.75,419717.41,-3415.66
100,1601827.44,1550369.92,51457.52
4884,1160744.12,1268766.76,-108022.64


In [16]:
mae = mean_absolute_error(y_test, y_pred) # computing mean absolute error
print(f'MAE: {mae}') 

MAE: 78686.98662004662


In [17]:
rmse = root_mean_squared_error(y_test, y_pred)
print(f'RMSE: {rmse}') 

RMSE: 144848.99461481292


In [18]:
r2_score = r2_score(y_test, y_pred)
print(f'R2_Score: {r2_score}') 


R2_Score: 0.931957923617315


- model fits the data well
- predictions are close to actual values
- model explains about 93.2% of the variance in the target variable.

# Deployment


In [19]:
import joblib

In [20]:
joblib.dump(model, 'model.pkl')

['model.pkl']

In [21]:
X_train.columns # features used for prediction

Index(['Store', 'Holiday_Flag', 'Temperature', 'Fuel_Price', 'CPI',
       'Unemployment'],
      dtype='str')